### PR TITLE
Check bounds for indexing tuples by logical masks

### LINKED ISSUE #19719: Boolean mask indexing of a tuple
Boolean indexing of tuples doesn't check bounds, but boolean indexing of vectors does.

```julia
julia> (1,2,3)[[true, false]]
(1,)

julia> (1,2,3)[[true, true, false, false, false]]
(1,2)

julia> [1,2,3][[true, false]]
ERROR: BoundsError: attempt to access 3-element Array{Int64,1} at index [Bool[true,false]]
 in throw_boundserror(::Array{Int64,1}, ::Tuple{Array{Bool,1}}) at ./abstractarray.jl:355
 in checkbounds at ./abstractarray.jl:284 [inlined]
 in _getindex at ./multidimensional.jl:270 [inlined]
 in getindex(::Array{Int64,1}, ::Array{Bool,1}) at ./abstractarray.jl:752
```

I can see a parallel with tuple assignment, so perhaps this was intentional? Is there an upside to allowing that? I could not find any discussion of it.

### PR DESCRIPTION
**Pull request by mbauman**
Also deprecate indexing tuples by non-vectors since that should increase the
dimensionality of the tuple, but tuples are 1-d only structures.

Fix #19719

### CODE REVIEW

File: base/tuple.jl [hunk 1]
```diff
@@ -15,8 +15,8 @@ endof(t::Tuple) = length(t)
 size(t::Tuple, d) = d==1 ? length(t) : throw(ArgumentError("invalid tuple dimension $d"))
 getindex(t::Tuple, i::Int) = getfield(t, i)
 getindex(t::Tuple, i::Real) = getfield(t, convert(Int, i))
-getindex(t::Tuple, r::AbstractArray) = tuple([t[ri] for ri in r]...)
-getindex(t::Tuple, b::AbstractArray{Bool}) = getindex(t,find(b))
+getindex{T}(t::Tuple, r::AbstractArray{T,1}) = tuple([t[ri] for ri in r]...)
+getindex(t::Tuple, b::AbstractArray{Bool,1}) = length(b) == length(t) ? getindex(t,find(b)) : throw(BoundsError(t, b))
```
Comment (tkelman) [inline base/tuple.jl]: apparently it's never size-checked for as long as this getindex signature has existed https://github.com/JuliaLang/julia/commit/66ae77a3d5f1c6cf607336497598bab9608c9125, wonder if there was a reason not to

**Comment by tkelman** [comment]
just in case this might somehow be a hotspot? @nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: base/tuple.jl (hunk 1 above)
Comment (mbauman) [inline base/tuple.jl]: Eh, I'd bet it was just an oversight. Bounds checking back then was [less picky](https://github.com/JuliaLang/julia/blob/66ae77a3d5f1c6cf607336497598bab9608c9125/base/abstractarray.jl#L59) and no tests broke with this change.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19737#issuecomment-269410832) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4b8380c639229b2a2837b3cd93825d54dffb39ff/cf83f45_vs_8ef3daa/report.md). cc @jrevels

**Review by tkelman — APPROVED** [review]
benchmark results are noise, lgtm

File: base/tuple.jl [hunk 2]
```diff
@@ -15,8 +15,8 @@ endof(t::Tuple) = length(t)
 size(t::Tuple, d) = d==1 ? length(t) : throw(ArgumentError("invalid tuple dimension $d"))
 getindex(t::Tuple, i::Int) = getfield(t, i)
 getindex(t::Tuple, i::Real) = getfield(t, convert(Int, i))
-getindex(t::Tuple, r::AbstractArray) = tuple([t[ri] for ri in r]...)
-getindex(t::Tuple, b::AbstractArray{Bool}) = getindex(t,find(b))
+getindex{T}(t::Tuple, r::AbstractArray{T,1}) = tuple([t[ri] for ri in r]...)
```
Comment (JeffBezanson) [inline base/tuple.jl]: `r::AbstractVector` ?

Comment (mbauman) [inline base/tuple.jl]: It's not defined at this point in bootstrap.

**Comment by JeffBezanson** [comment]
Lots of PRs adding stuff at the end of `deprecated.jl`. Frustrating :(

File: base/tuple.jl (hunk 2 above)
Comment (tkelman) [inline base/tuple.jl]: be sure to squash on merge then, if the first commit wouldn't build

**Pull request merged**